src/core/eme/index.ts

```diff
@@ -60,63 +61,146 @@ function createMediaKeysObs(
-        trySettingServerCertificate(
-          mediaKeys, serverCertificate, errorStream) :
-      Observable.empty()) as Observable<never>; // Typescript hack
-
-    const mksConfig = keySystemAccess.getConfiguration();
-
-    const setMediaKeys$ = setMediaKeysObs(
-      mediaKeys, mksConfig, video, keySystem, instanceInfos);
-    if (encryptedEvent.initData) {
-      const initData = new Uint8Array(encryptedEvent.initData);
-      const manageSessionCreation$ = manageSessionCreation(
-        mediaKeys, mksConfig, keySystem, encryptedEvent.initDataType,
-        initData, errorStream);
-
-      return setCertificate$
-        .concat(Observable.merge(setMediaKeys$, manageSessionCreation$));
-    } else {
-      const error = new Error("no init data found on media encrypted event.");
-      throw new EncryptedMediaError("INVALID_ENCRYPTED_EVENT", error, true);
-    }
+  const { keySystem, keySystemAccess } = keySystemInfo;
+  const { serverCertificate } = keySystem;
+  handleSessionStorage(keySystem);
+  const initData = new Uint8Array(encryptedEvent.initData);
+  const mksConfig = keySystemAccess.getConfiguration();
+
+  return mediaKeys$.mergeMap((mediaKeys) => {
+    const setMediaKeys$ =
+      setMediaKeysObs(mediaKeys, mksConfig, video, keySystem, instanceInfos);
+    const manageSessionCreation$ = manageSessionCreation(
+      mediaKeys, mksConfig, keySystem, encryptedEvent.initDataType,
+      initData, errorStream);
+    return setCertificate$(serverCertificate, mediaKeys, errorStream)
+      .concat(Observable.merge(setMediaKeys$, manageSessionCreation$));
   });
 }
 
+/**
+ * Function triggered when both:
+ *   - first event from current encrypted content has been received.
+ *   - a compatible key system configuration has been found.
+ *
+ * Manage session creation for current mediaKeys, through all subsequent EME APIs.
+ */
+function handleFirstPlaybackEncryptedEvents(
+  encryptedEvent : MediaEncryptedEvent,
+  keySystemInfo: IKeySystemPackage,
+  video : HTMLMediaElement,
+  errorStream: ErrorStream
+): Observable<MediaKeys|ISessionEvent|Event> {
+
+  const { keySystemAccess } = keySystemInfo;
+
+  return handleEncryptedEvent(
+    encryptedEvent,
+    keySystemInfo,
+    video,
+    errorStream,
+    createMediaKeysObs(keySystemAccess)
+  );
+}
+
+/**
+ * Function triggered when both:
+ *   - events from current encrypted content has been received.
+ *   - a compatible key system configuration has been found.
+ *
+ * Manage session creation for current mediaKeys, through all subsequent EME APIs.
+ */
+function handleOngoingPlaybackEncryptedEvents(
+  encryptedEvent : MediaEncryptedEvent,
+  keySystemInfo: IKeySystemPackage,
+  video : HTMLMediaElement,
+  errorStream: ErrorStream
+): Observable<MediaKeys|ISessionEvent|Event> {
+
+  if(video.mediaKeys == null){
+    const error = new Error("Video Element should have attached MediaKeys.");
+    throw new EncryptedMediaError("INVALID_ENCRYPTED_EVENT", error, true);
```

Comment:
That's not an ``"INVALID_ENCRYPTED_EVENT"``...

Plus, wouldn't it be theorically possible to get two really fast "encrypted" event and not have the time to set the MediaKeys yet?